integrations/sharepoint/src/SharepointClient.ts

```diff
@@ -0,0 +1,441 @@
+    const res = await axios.get<SharePointItemsResponse>(url, {
+      headers: {
+        Authorization: `Bearer ${token.accessToken}`,
+        Accept: "application/json;odata=verbose",
+      },
+    });
+    return res.data.d.results;
+  }
+
+  /**
+   * Get the file name from the SharePoint list item
+   * @param listItemIndex - The item index to get the file name for
+   * @returns - The file name or null if the file does not exist
+   */
+  private async getFileName(listItemIndex: number): Promise<string | null> {
+    // sample url -https://botpressio836.sharepoint.com/sites/DemoStandardTeamPage/_api/web/lists/getbytitle('NewDL')/items(3)/File
+    const url = `https://${this.primaryDomain}.sharepoint.com/sites/${this.siteName}/_api/web/lists/getbytitle('${this.documentLibraryName}')/items(${listItemIndex})/File`;
+    const token = await this.acquireToken();
+    const res = await axios
+      .get(url, {
+        headers: {
+          Authorization: `Bearer ${token.accessToken}`,
+          Accept: "application/json;odata=verbose",
+        },
+      })
+      .catch(() => {
+        return {
+          data: {
+            d: {
+              Name: null,
+            },
+          },
+        };
+      });
+    return res.data.d.Name;
+  }
+
+  /**
+   * A method to get changes from a SharePoint list
+   * @returns - The list of changes
+   */
+  private async getChanges(changeToken: string | null): Promise<ChangeItem[]> {
+    const token = await this.acquireToken();
+    const url = `https://${this.primaryDomain}.sharepoint.com/sites/${this.siteName}/_api/web/lists/getbytitle('${this.documentLibraryName}')/GetChanges`;
+
+    type GetChangesPayload = {
+      query: {
+        Item: boolean;
+        Add: boolean;
+        Update: boolean;
+        DeleteObject: boolean;
+        ChangeTokenStart?: {
+          StringValue: string;
+        };
+      };
+    };
+
+    const payload: GetChangesPayload = {
+      query: {
+        Item: true,
+        Add: true,
+        Update: true,
+        DeleteObject: true,
+      },
+    };
+
+    if (changeToken !== null) {
+      payload.query.ChangeTokenStart = {
+        StringValue: changeToken,
+      };
+    }
+
+    const res = await axios
+      .post<ChangeResponse>(url, payload, {
+        headers: {
+          Authorization: `Bearer ${token.accessToken}`,
+          Accept: "application/json;odata=verbose",
+        },
+      })
+      .catch(handleAxiosError);
+
+    if (!res) {
+      throw new sdk.RuntimeError(`Error getting changes`);
+    }
+
+    return res.data.d.results;
+  }
+
+  private log(message: string): void {
+    console.log(`[${getFormatedCurrTime()} - SP Client] ${message}`);
+  }
+
+  /**
+   * Processes changes from a SharePoint list
+   */
+  async syncSharepointDocumentLibraryAndBotpressKB(changeToken: string): Promise<string> {
+    const changes = await this.getChanges(changeToken);
+
+    const latestChange = changes.at(-1);
```

Comment:
**chore:** You have the same logic on line 117. You can move it into `getChanges` method

Reply:
Not sure if If I  fully understand what you mean here.. getChanges returns a array of changes. how can I move this there?

Reply:
Here are 2 spots where `getChanges` is called:
Line 114:
```
  const changes = await this.getChanges(null);
    if (changes.length > 0) {
      const latestChange = changes.at(-1);
      if (!latestChange) {
        throw new sdk.RuntimeError(`Error initializing change token`);
      }
      return latestChange.ChangeToken.StringValue;
    }
```
Line 330:
```
    const changes = await this.getChanges(changeToken);

    const latestChange = changes.at(-1);
    if (!latestChange) {
      console.log("No changes to process");
      return changeToken;
    }
```

Would it make sense to move the empty changes check to the `getChanges` method to avoid logic duplication

Reply:
umm not entirely. because in the first case it exists just for typescript to not throw a error. and and in the second case its actually is a valid state that we need to handle by returning the prev change token. 

maybe a `changes.length === 0` would make more sense, ill update the code.

Reply:
I also removed. the check for typescript, and replaced it with a non null assertion. should be clearer now! let me know what you think.